File: .taira-source-c80bf4d85a97/java/iroha_android/src/main/java/org/hyperledger/iroha/android/telemetry/TelemetryExportStatusSink.java
package org.hyperledger.iroha.android.telemetry;

import java.util.Map;
import java.util.Objects;
import org.hyperledger.iroha.android.client.ClientResponse;

/**
 * {@link TelemetrySink} wrapper that emits {@code android.telemetry.export.status} for every
 * delegate invocation. The wrapper records {@code status="ok"} when a delegate call succeeds and
 * {@code status="error"} when the delegate throws.
 */
public final class TelemetryExportStatusSink implements TelemetrySink {

  private static final String SIGNAL_ID = "android.telemetry.export.status";

  private final TelemetrySink delegate;
  private final String exporterName;

  private TelemetryExportStatusSink(final TelemetrySink delegate, final String exporterName) {
    this.delegate = delegate;
    this.exporterName = exporterName;
  }

  public static TelemetrySink wrap(final TelemetrySink delegate, final String exporterName) {
    if (delegate == null) {
      return null;
    }
    if (delegate instanceof TelemetryExportStatusSink) {
      return delegate;
    }
    final String resolvedName = resolveExporterName(delegate, exporterName);
    return new TelemetryExportStatusSink(delegate, resolvedName);
  }

  public static TelemetrySink unwrap(final TelemetrySink sink) {
    if (sink instanceof TelemetryExportStatusSink statusSink) {
      return statusSink.delegate;
    }
    return sink;
  }

  static String exporterName(final TelemetrySink sink) {
    if (sink instanceof TelemetryExportStatusSink statusSink) {
      return statusSink.exporterName;
    }
    return null;
  }

  private static String resolveExporterName(
      final TelemetrySink delegate, final String exporterName) {
    final String candidate = exporterName == null ? "" : exporterName.trim();
    if (!candidate.isEmpty()) {
      return candidate;
    }
    final String simpleName = delegate.getClass().getSimpleName();
    return simpleName == null || simpleName.isEmpty() ? "android_sdk" : simpleName;
  }

  @Override
  public void onRequest(final TelemetryRecord record) {
    invokeWithStatus(() -> delegate.onRequest(record));
  }

  @Override
  public void onResponse(final TelemetryRecord record, final ClientResponse response) {
    invokeWithStatus(() -> delegate.onResponse(record, response));
  }

  @Override
  public void onFailure(final TelemetryRecord record, final Throwable error) {
    invokeWithStatus(() -> delegate.onFailure(record, error));
  }

  @Override
  public void emitSignal(final String signalId, final Map<String, Object> fields) {
    if (Objects.equals(SIGNAL_ID, signalId)) {
      delegate.emitSignal(signalId, fields);
      return;
    }
    invokeWithStatus(() -> delegate.emitSignal(signalId, fields));
  }

  private void invokeWithStatus(final Runnable runnable) {
    try {
      runnable.run();
      emitStatus("ok");
    } catch (final RuntimeException ex) {
      emitStatus("error");
      throw ex;
    }
  }

  private void emitStatus(final String status) {
    try {
      delegate.emitSignal(
          SIGNAL_ID,
          Map.of(
              "exporter", exporterName,
              "status", status));
    } catch (final RuntimeException ignored) {
      // Best-effort metrics; swallow exporter failures to avoid infinite recursion.
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/telemetry/TelemetryExportStatusSink.kt
package org.hyperledger.iroha.sdk.telemetry

import org.hyperledger.iroha.sdk.client.ClientResponse

private const val SIGNAL_ID = "android.telemetry.export.status"

/**
 * `TelemetrySink` wrapper that emits `android.telemetry.export.status` for every
 * delegate invocation. The wrapper records `status="ok"` when a delegate call succeeds and
 * `status="error"` when the delegate throws.
 */
class TelemetryExportStatusSink private constructor(
    private val delegate: TelemetrySink,
    private val exporterName: String,
) : TelemetrySink {

    override fun onRequest(record: TelemetryRecord) {
        invokeWithStatus { delegate.onRequest(record) }
    }

    override fun onResponse(record: TelemetryRecord, response: ClientResponse) {
        invokeWithStatus { delegate.onResponse(record, response) }
    }

    override fun onFailure(record: TelemetryRecord, error: Throwable) {
        invokeWithStatus { delegate.onFailure(record, error) }
    }

    override fun emitSignal(signalId: String, fields: Map<String, Any>) {
        if (SIGNAL_ID == signalId) {
            delegate.emitSignal(signalId, fields)
            return
        }
        invokeWithStatus { delegate.emitSignal(signalId, fields) }
    }

    private fun invokeWithStatus(action: Runnable) {
        try {
            action.run()
            emitStatus("ok")
        } catch (ex: RuntimeException) {
            emitStatus("error")
            throw ex
        }
    }

    private fun emitStatus(status: String) {
        try {
            delegate.emitSignal(
                SIGNAL_ID,
                mapOf(
                    "exporter" to exporterName,
                    "status" to status,
                ),
            )
        } catch (_: RuntimeException) {
            // Best-effort metrics; swallow exporter failures to avoid infinite recursion.
        }
    }

    companion object {
        @JvmStatic
        fun wrap(delegate: TelemetrySink?, exporterName: String?): TelemetrySink? {
            if (delegate == null) return null
            if (delegate is TelemetryExportStatusSink) return delegate
            val resolvedName = resolveExporterName(delegate, exporterName)
            return TelemetryExportStatusSink(delegate, resolvedName)
        }

        @JvmStatic
        fun unwrap(sink: TelemetrySink?): TelemetrySink? =
            if (sink is TelemetryExportStatusSink) sink.delegate else sink

        @JvmStatic
        fun exporterName(sink: TelemetrySink): String? =
            if (sink is TelemetryExportStatusSink) sink.exporterName else null

        private fun resolveExporterName(delegate: TelemetrySink, exporterName: String?): String {
            val candidate = exporterName?.trim() ?: ""
            if (candidate.isNotEmpty()) return candidate
            val simpleName = delegate.javaClass.simpleName
            return if (simpleName.isNullOrEmpty()) "android_sdk" else simpleName
        }
    }
}
